src/data_structures/fmindex.rs

```diff
@@ -88,41 +100,78 @@ pub trait FMIndexable {
 
 }
 
-/// The Fast Index in Minute space (FM-Index, Ferragina and Manzini, 2000) for finding suffix array
-/// intervals matching a given pattern.
+pub trait FMIndexCore {
+    type SA: SuffixArray;
 
-#[cfg_attr(feature = "serde_macros", derive(Serialize, Deserialize))]
-pub struct FMIndex<
-        DBWT: DerefBWT + Clone,
-        DLess: DerefLess + Clone,
-        DOcc: DerefOcc + Clone> {
-    bwt: DBWT,
-    less: DLess,
-    occ: DOcc,
+    fn occ(&self, r: usize, a: u8) -> usize;
+    fn less(&self, a: u8) -> usize;
+    fn bwt(&self) -> &BWT;
+    fn sa(&self) -> &Self::SA;
 }
 
-impl<
-    DBWT: DerefBWT + Clone,
-    DLess: DerefLess + Clone,
-    DOcc: DerefOcc + Clone> FMIndexable for FMIndex<DBWT, DLess, DOcc> {
+impl FMIndexCore for SampledSuffixArray {
+    // the trait needs to track what kind of suffix array reference it's returning
+    type SA = SampledSuffixArray;
```

Comment:
> the FM index doesn't depend on the suffix array

But the FM Index search returns intervals of the suffix array -- there's not much use to being able to search an FM index without a suffix array to retrieve the positions of the pattern in the original text. Unless I'm missing something?